### PR TITLE
AC-181: Dynamic Form Creation from JSON

### PR DESCRIPTION
**Pull request by AvijitGhosh82**

### CODE REVIEW

File: openmrs-client/src/main/java/org/openmrs/mobile/activities/FormCreateActivity.java [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+        btnNext.setOnClickListener(this);
+        btnFinish.setOnClickListener(this);
+
+
+        String formjson=loadAssetTextAsString(FormCreateActivity.this,"openmrs-forms/vitals.json");
+        //Toast.makeText(this,formjson,Toast.LENGTH_SHORT).show();
+
+        Gson gson = new Gson();
+        form=gson.fromJson(formjson,Form.class);
+        pagelist = form.getPages();
```
Comment (rkorytkowski) [inline openmrs-client/src/main/java/org/openmrs/mobile/activities/FormCreateActivity.java]: Please move lines 78-83 to a dedicated class. I should be able to call. FormService formService = new FormService(); form = formService.getFormByName("Capture Vitals");

File: openmrs-client/src/main/java/org/openmrs/mobile/activities/FormCreateActivity.java [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+
+        btnNext = (Button) findViewById(R.id.btn_next);
+        btnFinish = (Button) findViewById(R.id.btn_finish);
+        pager_indicator = (LinearLayout) findViewById(R.id.viewPagerCountDots);
+
+        btnNext.setOnClickListener(this);
+        btnFinish.setOnClickListener(this);
+
+
+        String formjson=loadAssetTextAsString(FormCreateActivity.this,"openmrs-forms/vitals.json");
```
Comment (rkorytkowski) [inline openmrs-client/src/main/java/org/openmrs/mobile/activities/FormCreateActivity.java]: Can't you just use JsonReader instead?

File: openmrs-client/src/main/java/org/openmrs/mobile/activities/FormCreateActivity.java [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+                } catch (IOException e) {
+                    //Log.e(TAG, "Error closing asset " + name);
+                }
+            }
+        }
+
+        return null;
+    }
+
+    private void setUiPageViewController() {
```
Comment (rkorytkowski) [inline openmrs-client/src/main/java/org/openmrs/mobile/activities/FormCreateActivity.java]: What does it exactly do?

File: openmrs-client/src/main/java/org/openmrs/mobile/activities/FormCreateActivity.java [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+        }
+    }
+
+    @Override
+    public void onPageScrollStateChanged(int state) {
+
+    }
+
+
+    public class SectionsPagerAdapter extends FragmentPagerAdapter {
```
Comment (rkorytkowski) [inline openmrs-client/src/main/java/org/openmrs/mobile/activities/FormCreateActivity.java]: My understanding is that this class is responsible for displaying a form. How about we extract it to a separate file and change the constructor to SectionsPageAdapter(FragmentManager fm, Form form) and store the form inside it. I'd also rename it to FormUiBuilder.

File: openmrs-client/src/main/java/org/openmrs/mobile/activities/FormCreateActivity.java [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+import org.openmrs.mobile.models.retrofit.form.Form;
+import org.openmrs.mobile.models.retrofit.form.Page;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.util.List;
+
+public class FormCreateActivity extends ACBaseActivity implements ViewPager.OnPageChangeListener,View.OnClickListener{
```
Comment (rkorytkowski) [inline openmrs-client/src/main/java/org/openmrs/mobile/activities/FormCreateActivity.java]: Rename to FormDisplayActivity

File: openmrs-client/src/main/java/org/openmrs/mobile/activities/FormCreateActivity.java [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+
+    private SectionsPagerAdapter mSectionsPagerAdapter;
+
+
+    private ViewPager mViewPager;
+    private Button btnNext, btnFinish;
+    private LinearLayout pager_indicator;
+    private int dotsCount;
+    private ImageView[] dots;
+    private Form form;
```
Comment (rkorytkowski) [inline openmrs-client/src/main/java/org/openmrs/mobile/activities/FormCreateActivity.java]: No need to sore form in here, please see my comment on SectionsPagerAdapter

File: openmrs-client/src/main/java/org/openmrs/mobile/activities/FormCreateActivity.java [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+    List<Page> pagelist;
```
Comment (rkorytkowski) [inline openmrs-client/src/main/java/org/openmrs/mobile/activities/FormCreateActivity.java]: No need to store pagelist here, please see my comment on SectionsPagerAdapter

File: openmrs-client/src/main/java/org/openmrs/mobile/activities/FormCreateActivity.java (hunk 4 above)
Comment (rkorytkowski) [inline openmrs-client/src/main/java/org/openmrs/mobile/activities/FormCreateActivity.java]: Can we actually do more refactoring and have FormDisplayGenerator be a fragment, which contains next, finish and all logic for moving between pages and contain SectionsPagerAdapter renamed to FormPagerAdapter?

Comment (AvijitGhosh82) [inline openmrs-client/src/main/java/org/openmrs/mobile/activities/FormCreateActivity.java]: Should the dots logic not be a part of the main activity? Each page is a fragment..so the dots (indicators) should be in the parent activity. No?

**Comment by dkayiwa** [comment]
These look like requiring a squashing into one commit. :)

**Comment by AvijitGhosh82** [comment]
@rkorytkowski Observations are being created properly, the Encounter is giving a 500 error. I have posted details on Talk. Once that is fixed this looks done, the VisitDAO and EncounterDAO can be used to save the data locally.

**Comment by AvijitGhosh82** [comment]
@rkorytkowski I fixed it. Everything is working. :)

**Comment by dkayiwa** [comment]
@AvijitGhosh82 can you squash the commits into one?

**Pull request closed without merging**